libs/cart/src/reducers/cart-shipping-address/cart-shipping-address.reducer.ts

```diff
@@ -0,0 +1,48 @@
+import {
+  DaffCartShippingAddressActionTypes,
+} from '../../actions';
+import { initialState } from '../cart-initial-state';
+import { CartState } from '../cart-state.interface';
+import { ActionTypes } from '../action-types.type';
+
+export function reducer(
+  state = initialState,
+  action: ActionTypes
+): CartState {
+  switch (action.type) {
+    case DaffCartShippingAddressActionTypes.CartShippingAddressLoadAction:
+    case DaffCartShippingAddressActionTypes.CartShippingAddressUpdateAction:
+      return { ...state, loading: true };
+
+    case DaffCartShippingAddressActionTypes.CartShippingAddressLoadSuccessAction:
+      return {
+        ...state,
+        cart: {
+          ...state.cart,
+          shipping_address: action.payload
+        },
+        loading: false
```

Comment:
Maybe success actions should reset the errors to []? What do you think?

Reply:
The errors are shared by all the different driver actions so we might be clearing errors that are still relevant.

Reply:
Good point. I think we need a more nuanced errors object then, because if we don't clear the errors on success, we won't know when there's really an error or not. I think a good solution could something like: 
```
export interface DaffCartReducerState {
  cart: DaffCart,
  loading: boolean,
  errors: DaffCartError[]
}

export interface DaffCartError {
   type: DaffCartErrorTypeEnum;
   errors: string[];
}

export enum DaffCartErrorTypeEnum {
   ShippingAddress = "Shipping Address",
   Cart = "Cart",
   CartItem = "Cart Item",
   ShippingMethod = "Shipping Method",
   ...
}
```
then in the reducers do:
```
export function reducer(
  state = initialState,
  action: ActionTypes
): DaffCartReducerState {
  switch (action.type) {
    case DaffCartActionTypes.CartCreateSuccessAction:
      return {
        ...state,
        cart: {
          ...state.cart,
          ...action.payload
        },
        loading: false,
        errors: state.errors.filter(error => error.type !== DaffCartErrorTypeEnum.Cart)
      };

    case DaffCartActionTypes.CartCreateFailureAction:
      return {
        ...state,
        loading: false,
        errors: state.errors.concat({ type: DaffCartErrorTypeEnum.Cart, error: action.payload })
      };

    default:
      return state;
  }
}
```
Then people could filter on different feature areas and get specific errors. We could also make selectors for getting errors for specific feature areas. That'd be cool.

Reply:
What about storing the errors in an object keyed by the enum? i.e.
```ts
export interface DaffCartReducerState {
  cart: DaffCart,
  loading: boolean,
  errors: DaffCartErrors
}

export interface DaffCartErrors {
   [key: DaffCartErrorTypeEnum]: string[]
}
```

It would make selecting for types of errors easier and more performant.

Reply:
sure, that seems better

Reply:
Changed in [`42f64c5`](https://github.com/graycoreio/daffodil/pull/681/commits/42f64c5c33b8d2ce7071d15a299d7cab2ef0e23f).

Reply:
Changed to do error reducing logic in separate methods in [`d8b7537`](https://github.com/graycoreio/daffodil/pull/681/commits/d8b75371b3f15b19a655a63800913bb975dbb614).